Bottom: 68569a4fcc9a67d8cf78103f37c5c51a28c60f7c
Top:    09cd1a18eb13f0784bcf924c068c90413a63a619
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2009-10-21 00:41:33 +0100

Add infrastructure for subcompositing various bits of our drawing














---

diff --git a/src/draw.c b/src/draw.c
index 1d02376..b75989a 100644
--- a/src/draw.c
+++ b/src/draw.c
@@ -354,7 +354,7 @@ static int
 lowvia_callback (const BoxType * b, void *cl)
 {
   PinTypePtr via = (PinTypePtr) b;
-  if (!via->Mask)
+//  if (!via->Mask)
     DrawPlainVia (via, False);
   return 1;
 }
@@ -474,27 +474,9 @@ DrawEverything (BoxTypePtr drawn_area)
     }
   if (TEST_FLAG (CHECKPLANESFLAG, PCB) && gui->gui)
     return;
-
   /* draw vias below silk */
   if (PCB->ViaOn && gui->gui)
     r_search (PCB->Data->via_tree, drawn_area, NULL, lowvia_callback, NULL);
-  /* Draw the solder mask if turned on */
-  if (gui->set_layer ("componentmask", SL (MASK, TOP), 0))
-    {
-      int save_swap = SWAP_IDENT;
-      SWAP_IDENT = 0;
-      DrawMask (drawn_area);
-      SWAP_IDENT = save_swap;
-      gui->set_layer (NULL, SL (FINISHED, 0), 0);
-    }
-  if (gui->set_layer ("soldermask", SL (MASK, BOTTOM), 0))
-    {
-      int save_swap = SWAP_IDENT;
-      SWAP_IDENT = 1;
-      DrawMask (drawn_area);
-      SWAP_IDENT = save_swap;
-      gui->set_layer (NULL, SL (FINISHED, 0), 0);
-    }
   /* Draw pins, pads, vias below silk */
   if (gui->gui)
     DrawTop (drawn_area);
@@ -523,6 +505,23 @@ DrawEverything (BoxTypePtr drawn_area)
 		    &plated);
 	}
     }
+  /* Draw the solder mask if turned on */
+  if (gui->set_layer ("componentmask", SL (MASK, TOP), 0))
+    {
+      int save_swap = SWAP_IDENT;
+      SWAP_IDENT = 0;
+      DrawMask (drawn_area);
+      SWAP_IDENT = save_swap;
+      gui->set_layer (NULL, SL (FINISHED, 0), 0);
+    }
+  if (gui->set_layer ("soldermask", SL (MASK, BOTTOM), 0))
+    {
+      int save_swap = SWAP_IDENT;
+      SWAP_IDENT = 1;
+      DrawMask (drawn_area);
+      SWAP_IDENT = save_swap;
+      gui->set_layer (NULL, SL (FINISHED, 0), 0);
+    }
   /* Draw top silkscreen */
   if (gui->set_layer ("topsilk", SL (SILK, TOP), 0))
     {
@@ -543,7 +542,7 @@ DrawEverything (BoxTypePtr drawn_area)
 	r_search (PCB->Data->element_tree, drawn_area, NULL, EMark_callback,
 		  NULL);
       /* Draw rat lines on top */
-      if (PCB->RatOn)
+      if (PCB->RatOn && gui->set_layer ("rats", SL (RATS, 0), 0))
 	DrawRats(drawn_area);
     }
 
@@ -643,7 +642,7 @@ static int
 via_callback (const BoxType * b, void *cl)
 {
   PinTypePtr via = (PinTypePtr) b;
-  if (via->Mask)
+//  if (via->Mask)
     DrawPlainVia (via, False);
   return 1;
 }
@@ -720,6 +719,8 @@ clearPad_callback (const BoxType * b, void *cl)
   return 1;
 }
 
+#define MASK_SUBCOMPOSITE_SILK 0
+
 /* ---------------------------------------------------------------------------
  * Draws silk layer.
  */
@@ -742,11 +743,27 @@ DrawSilk (int new_swap, int layer, BoxTypePtr drawn_area)
       gui->use_mask (HID_MASK_BEFORE);
 #endif
       DrawLayer (LAYER_PTR (max_layer + layer), drawn_area);
+
+#if MASK_SUBCOMPOSITE_SILK
+      gui->use_mask (HID_MASK_BEFORE);
+      gui->use_mask (HID_MASK_CLEAR);
+      gui->fill_rect (Output.fgGC, 0, 0, PCB->MaxWidth, PCB->MaxHeight);
+      gui->use_mask (HID_MASK_BEFORE);
+#endif
+
       /* draw package */
       r_search (PCB->Data->element_tree, drawn_area, NULL, frontE_callback,
 		NULL);
       r_search (PCB->Data->name_tree[NAME_INDEX (PCB)], drawn_area, NULL,
 		frontN_callback, NULL);
+
+#if MASK_SUBCOMPOSITE_SILK
+      gui->use_mask (HID_MASK_AFTER);
+      gui->set_color (Output.fgGC, PCB->ElementColor);
+      gui->fill_rect (Output.fgGC, 0, 0, PCB->MaxWidth, PCB->MaxHeight);
+      gui->use_mask (HID_MASK_OFF);
+#endif
+
 #if 0
     }
 
@@ -869,6 +886,7 @@ text_callback (const BoxType * b, void *cl)
   return 1;
 }
 
+#define MASK_SUBCOMPOSITE_LAYERS 0
 
 /* ---------------------------------------------------------------------------
  * draws one non-copper layer
@@ -878,6 +896,13 @@ DrawLayer (LayerTypePtr Layer, BoxType * screen)
 {
   struct pin_info info;
 
+#if MASK_SUBCOMPOSITE_LAYERS
+  gui->use_mask (HID_MASK_BEFORE);
+  gui->use_mask (HID_MASK_CLEAR);
+  gui->fill_rect (Output.fgGC, 0, 0, PCB->MaxWidth, PCB->MaxHeight);
+  gui->use_mask (HID_MASK_BEFORE);
+#endif
+
   /* print the non-clearing polys */
   info.Layer = Layer;
   info.arg = False;
@@ -893,8 +918,17 @@ DrawLayer (LayerTypePtr Layer, BoxType * screen)
   /* draw the layer text on screen */
   r_search (Layer->text_tree, screen, NULL, text_callback, Layer);
   clip_box = NULL;
+
+#if MASK_SUBCOMPOSITE_LAYERS
+  gui->use_mask (HID_MASK_AFTER);
+  gui->set_color (Output.fgGC, Layer->Color);
+  gui->fill_rect (Output.fgGC, 0, 0, PCB->MaxWidth, PCB->MaxHeight);
+  gui->use_mask (HID_MASK_OFF);
+#endif
 }
 
+#define MASK_SUBCOMPOSITE_LAYER_GROUPS 0
+
 /* ---------------------------------------------------------------------------
  * draws one layer group.  Returns non-zero if pins and pads should be
  * drawn with this group.
@@ -920,6 +954,14 @@ DrawLayerGroup (int group, const BoxType * screen)
 	rv = 0;
       if (layernum < max_layer && Layer->On)
 	{
+
+#if MASK_SUBCOMPOSITE_LAYER_GROUPS
+          gui->use_mask (HID_MASK_BEFORE);
+          gui->use_mask (HID_MASK_CLEAR);
+          gui->fill_rect (Output.fgGC, 0, 0, PCB->MaxWidth, PCB->MaxHeight);
+          gui->use_mask (HID_MASK_BEFORE);
+#endif
+
 	  /* draw all polygons on this layer */
 	  if (Layer->PolygonN)
 	    {
@@ -946,6 +988,12 @@ DrawLayerGroup (int group, const BoxType * screen)
 	  /* draw the layer text on screen */
 	  r_search (Layer->text_tree, screen, NULL, text_callback, Layer);
 
+#if MASK_SUBCOMPOSITE_LAYER_GROUPS
+          gui->use_mask (HID_MASK_AFTER);
+          gui->set_color (Output.fgGC, Layer->Color);
+          gui->fill_rect (Output.fgGC, 0, 0, PCB->MaxWidth, PCB->MaxHeight);
+          gui->use_mask (HID_MASK_OFF);
+#endif
 	}
     }
   if (n_entries > 1)
@@ -2159,6 +2207,8 @@ DrawElement (ElementTypePtr Element, int unused)
   DrawElementPinsAndPads (Element, unused);
 }
 
+#define MASK_SUBCOMPOSITE_ELEMENT_NAMES 0
+
 /* ---------------------------------------------------------------------------
  * draws the name of an element
  */
@@ -2169,6 +2219,15 @@ DrawElementName (ElementTypePtr Element, int unused)
     return;
   if (TEST_FLAG (HIDENAMEFLAG, Element))
     return;
+#if MASK_SUBCOMPOSITE_ELEMENT_NAMES
+  gui->use_mask (HID_MASK_BEFORE);
+  gui->use_mask (HID_MASK_CLEAR);
+  gui->fill_rect (Output.fgGC, 0, 0, PCB->MaxWidth, PCB->MaxHeight);
+  gui->use_mask (HID_MASK_BEFORE);
+
+  DrawTextLowLevel (&ELEMENT_TEXT (PCB, Element), PCB->minSlk);
+#endif
+
   if (doing_pinout || doing_assy)
     gui->set_color (Output.fgGC, PCB->ElementColor);
   else if (TEST_FLAG (SELECTEDFLAG, &ELEMENT_TEXT (PCB, Element)))
@@ -2177,15 +2236,33 @@ DrawElementName (ElementTypePtr Element, int unused)
     gui->set_color (Output.fgGC, PCB->ElementColor);
   else
     gui->set_color (Output.fgGC, PCB->InvisibleObjectsColor);
+
+#if MASK_SUBCOMPOSITE_ELEMENT_NAMES
+  gui->use_mask (HID_MASK_AFTER);
+  gui->fill_rect (Output.fgGC, 0, 0, PCB->MaxWidth, PCB->MaxHeight);
+  gui->use_mask (HID_MASK_OFF);
+#else
   DrawTextLowLevel (&ELEMENT_TEXT (PCB, Element), PCB->minSlk);
+#endif
 }
 
+#define MASK_SUBCOMPOSITE_ELEMENT_PACKAGE 0
+
 /* ---------------------------------------------------------------------------
  * draws the package of an element
  */
 void
 DrawElementPackage (ElementTypePtr Element, int unused)
 {
+#if MASK_SUBCOMPOSITE_ELEMENT_PACKAGE
+  gui->use_mask (HID_MASK_BEFORE);
+  gui->use_mask (HID_MASK_CLEAR);
+  gui->fill_rect (Output.fgGC, 0, 0, PCB->MaxWidth, PCB->MaxHeight);
+  gui->use_mask (HID_MASK_BEFORE);
+
+  DrawElementPackageLowLevel (Element, unused);
+#endif
+
   /* set color and draw lines, arcs, text and pins */
   if (doing_pinout || doing_assy)
     gui->set_color (Output.fgGC, PCB->ElementColor);
@@ -2195,7 +2272,14 @@ DrawElementPackage (ElementTypePtr Element, int unused)
     gui->set_color (Output.fgGC, PCB->ElementColor);
   else
     gui->set_color (Output.fgGC, PCB->InvisibleObjectsColor);
+
+#if MASK_SUBCOMPOSITE_ELEMENT_PACKAGE
+  gui->use_mask (HID_MASK_AFTER);
+  gui->fill_rect (Output.fgGC, 0, 0, PCB->MaxWidth, PCB->MaxHeight);
+  gui->use_mask (HID_MASK_OFF);
+#else
   DrawElementPackageLowLevel (Element, unused);
+#endif
 }
 
 /* ---------------------------------------------------------------------------
